### PR TITLE
refactor: unify template file processor

### PR DESCRIPTION
**Pull request by steinliber**
## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests

## Description
- this refactor will put all template file logic in `pkg/util/file`
- reposcaffolding will also use this logic later
- gitops has been tested

## New Behavior (screenshots if needed)

image

### CODE REVIEW

File: pkg/util/file/file.go [hunk 1]
```diff
@@ -1,14 +1,98 @@
...
+type fileProcesser func(string) (string, error)
+type renderProcesser func(string, string, map[string]interface{}) (string, error)
+
+// TemplateConfig contains all file template getter, process and render info
+type TemplateConfig struct {
+	info         string
+	templateName string
+	vars         map[string]interface{}
+	getter       fileProcesser
+	processer    fileProcesser
```
Comment (iyear) [inline pkg/util/file/file.go:24]: typo `processor`

File: pkg/util/file/file.go [hunk 2]
```diff
@@ -1,14 +1,98 @@
...
+	render       renderProcesser
+	processErr   error
+}
+
+func NewTempate() *TemplateConfig {
```
Comment (iyear) [inline pkg/util/file/file.go:29]: typo

File: pkg/util/file/file.go [hunk 3]
```diff
@@ -1,14 +1,98 @@
...
+	return c
+}
+
+func (c *TemplateConfig) Remote(url string) *TemplateConfig {
+	c.getter = getFileFromURL
+	c.info = url
+	return c
+}
+
+func (c *TemplateConfig) RenderFile(tempalateName string, vars map[string]interface{}) *TemplateConfig {
```
Comment (iyear) [inline pkg/util/file/file.go:51]: `tempalateName` -> `templateName`

File: pkg/util/file/file.go [hunk 4]
```diff
@@ -1,14 +1,98 @@
...
+
+func (c *TemplateConfig) Run() (string, error) {
+	// check if has error before
+	if c.processErr != nil {
+		return "", c.processErr
+	}
+
+	// check if info is empty
+	if c.info == "" {
+		return "", errors.New("File util: content is not setted")
```
Comment (iyear) [inline pkg/util/file/file.go:66]: Error string should not be capitalized or end with punctuation mark

File: pkg/util/file/file.go [hunk 5]
```diff
@@ -1,14 +1,98 @@
...
 
 import (
+	"errors"
 	"io/ioutil"
+	"os"
+	"path/filepath"
 )
 
+const (
+	defautlTempName    = "pkg-util-file-create_"
```
Comment (iyear) [inline pkg/util/file/file.go:11]: typo

File: pkg/util/file/file.go [hunk 6]
```diff
@@ -1,14 +1,98 @@
...
+	// check if info is empty
+	if c.info == "" {
+		return "", errors.New("File util: content is not setted")
+	}
+	var (
+		outPutName string
+		err        error
+	)
+	// 1. run Getter func to get file
+	outPutName, err = c.getter(c.info)
```
Comment (iyear) [inline pkg/util/file/file.go:73]: outPutName, err := c.getter(c.info)

File: pkg/util/file/file.go [hunk 7]
```diff
@@ -1,14 +1,98 @@
...
+	processer    fileProcesser
+	render       renderProcesser
+	processErr   error
+}
+
+func NewTempate() *TemplateConfig {
+	return &TemplateConfig{}
+}
+
+func (c *TemplateConfig) Local(path string) *TemplateConfig {
```
Comment (iyear) [inline pkg/util/file/file.go:33]: `FromLocal` is more clearer

File: pkg/util/file/file.go [hunk 8]
```diff
@@ -1,14 +1,98 @@
...
+		c.processErr = err
+	} else {
+		exPath := filepath.Dir(ex)
+		c.info = filepath.Join(exPath, path)
+	}
+	c.getter = getFileFromLocal
+	return c
+}
+
+func (c *TemplateConfig) Remote(url string) *TemplateConfig {
```
Comment (iyear) [inline pkg/util/file/file.go:45]: `FromRemote` is more clearer

File: internal/pkg/plugin/argocdapp/argocdapp.go [hunk 9]
```diff
@@ -1,27 +1,3 @@
...
-      - {{.source.valuefile}}
-    path: {{.source.path}}
-    repoURL: {{.source.repoURL}}
-    targetRevision: HEAD
-  syncPolicy:
-    automated:
-      prune: true
-      selfHeal: true
-`
+var templateFileLoc = "internal/pkg/plugin/argocdapp/tpl/app.tpl.yaml"
```
Comment (iyear) [inline internal/pkg/plugin/argocdapp/argocdapp.go:3]: use `go:embed` and add `FromReader` for `TemplateConfig`

Comment (iyear) [inline internal/pkg/plugin/argocdapp/argocdapp.go:3]: use `go:embed` and add `FromString` to `TemplateConfig`

**Review by iyear — APPROVED** [review]
Nice refactoring!

**Review by IronCore864 — APPROVED** [review]

**Pull request merged**